### PR TITLE
cockroachdb-bin: init at 23.1.7

### PR DESCRIPTION
**Pull request by neosimsim**
As part of #152626 we figured that recent versions of cockroach became nearly impossible to build.
As solution I suggest we use the pre-build releases in FHS.

Unfortunately version 20.1.8 has not been released for `aarch` so we have to drop these supported platforms, for now.

However I tested the derivation with a few newer release and they all worked, so updating cockroach should become a lot easier in the future.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module

### CODE REVIEW

**Comment by neosimsim** [comment]
@ofborg build cockroachdb

**Comment by neosimsim** [comment]
Is there a way to build unfree packages with ofborg?

File: pkgs/servers/sql/cockroachdb/default.nix [hunk 1]
```diff
@@ -1,70 +1,52 @@
...
+      sha256 = lib.fakeHash; # released for later versions only
+    };
+    x86_64-darwin = {
+      name = "cockroach-release-${version}";
+      url = "https://binaries.cockroachdb.com/cockroach-v${version}.darwin-10.9-amd64.tgz";
+      sha256 = "0k8q8aszik437x7if4y3z19rwdkb0mra9w5gip00v0y673hsajpf";
+    };
+    aarch64-darwin = {
+      name = "cockroach-release-${version}";
+      url = "https://binaries.cockroachdb.com/cockroach-v${version}.darwin-11.0-arm64.tgz";
```
Comment (turion) [inline pkgs/servers/sql/cockroachdb/default.nix:34]: Any reason why these are different darwin versions?

File: pkgs/servers/sql/cockroachdb/default.nix [hunk 2]
```diff
@@ -1,70 +1,52 @@
...
-  linuxDeps  = [ ncurses6 ];
-
-  buildInputs = if stdenv.isDarwin then darwinDeps else linuxDeps;
-  nativeBuildInputs = [ installShellFiles cmake xz which autoconf ];
-
-in
-buildGoPackage rec {
-  pname = "cockroach";
+  name = "cockroach";
   version = "20.1.8";
```
Comment (turion) [inline pkgs/servers/sql/cockroachdb/default.nix:8]: Isn't that a very old version? Or should we bump versions in a later PR?

File: pkgs/servers/sql/cockroachdb/default.nix [hunk 3]
```diff
@@ -1,70 +1,52 @@
...
-  env.NIX_CFLAGS_COMPILE = toString (lib.optionals stdenv.cc.isGNU [ "-Wno-error=deprecated-copy" "-Wno-error=redundant-move" "-Wno-error=pessimizing-move" ]);
-
-  inherit nativeBuildInputs buildInputs;
+  # For several reasons building cockroach from source has become
+  # nearly imposible. See https://github.com/NixOS/nixpkgs/pull/152626
+  # Therefore we use the pre-build release binary and wrap it with buildFHSUserEnv to
+  # work on nix.
+  # You can generate the hashes with
+  # nix-prefetch-url --type sha256 --unpack ...
+  src = fetchTarball (lib.getAttr stdenv.system {
```
Comment (wegank) [inline pkgs/servers/sql/cockroachdb/default.nix:16]: `fetchTarball` is forbidden in restricted mode. Use `fetchurl` instead.

File: pkgs/servers/sql/cockroachdb/default.nix [hunk 4]
```diff
@@ -1,70 +1,52 @@
...
-
-  env.NIX_CFLAGS_COMPILE = toString (lib.optionals stdenv.cc.isGNU [ "-Wno-error=deprecated-copy" "-Wno-error=redundant-move" "-Wno-error=pessimizing-move" ]);
-
-  inherit nativeBuildInputs buildInputs;
+  # For several reasons building cockroach from source has become
+  # nearly imposible. See https://github.com/NixOS/nixpkgs/pull/152626
+  # Therefore we use the pre-build release binary and wrap it with buildFHSUserEnv to
+  # work on nix.
+  # You can generate the hashes with
+  # nix-prefetch-url --type sha256 --unpack ...
```
Comment (wegank) [inline pkgs/servers/sql/cockroachdb/default.nix:15]:
```suggestion
  # nix store prefetch-file <url>
```

File: pkgs/servers/sql/cockroachdb/default.nix [hunk 5]
```diff
@@ -1,70 +1,52 @@
...
+  src = fetchTarball (lib.getAttr stdenv.system {
+    x86_64-linux = {
+      name = "cockroach-release-${version}";
+      url = "https://binaries.cockroachdb.com/cockroach-v${version}.linux-amd64.tgz";
+      sha256 = "1ra2l02psy449hay89nh87qwh3rspfcfb5i9r9gasd0qx8ybnbpc";
```
Comment (wegank) [inline pkgs/servers/sql/cockroachdb/default.nix:20]: SRI hashes, please.

File: pkgs/servers/sql/cockroachdb/default.nix (hunk 2 above)
Comment (neosimsim) [reply pkgs/servers/sql/cockroachdb/default.nix:8]: I would suggest du bump it later.

File: pkgs/servers/sql/cockroachdb/default.nix (hunk 1 above)
Comment (neosimsim) [reply pkgs/servers/sql/cockroachdb/default.nix:34]: I believe the M1 chip was introduces with macos 11.

File: pkgs/servers/sql/cockroachdb/default.nix (hunk 4 above)
Comment (neosimsim) [reply pkgs/servers/sql/cockroachdb/default.nix:15]: Thanks

File: pkgs/servers/sql/cockroachdb/default.nix (hunk 5 above)
Comment (neosimsim) [reply pkgs/servers/sql/cockroachdb/default.nix:20]: Now that I know of `store prefetch-file <url>`, with pleasure.

File: pkgs/servers/sql/cockroachdb/default.nix (hunk 3 above)
Comment (neosimsim) [reply pkgs/servers/sql/cockroachdb/default.nix]: Thanks

File: pkgs/servers/sql/cockroachdb/default.nix (hunk 1 above)
Comment (neosimsim) [reply pkgs/servers/sql/cockroachdb/default.nix]: I guess that's why they associate arm64 with darwin 11.

File: pkgs/servers/sql/cockroachdb/default.nix [hunk 6]
```diff
@@ -1,70 +1,52 @@
-{ lib, stdenv, buildGoPackage, fetchurl
-, cmake, xz, which, autoconf
-, ncurses6, libedit, libunwind
-, installShellFiles
-, removeReferencesTo, go
+{ lib
+, stdenv
```
Comment (wegank) [inline pkgs/servers/sql/cockroachdb/default.nix:2]:
```suggestion
, stdenv
, fetchurl
```

File: pkgs/servers/sql/cockroachdb/default.nix [hunk 7]
```diff
@@ -1,70 +1,52 @@
...
-  env.NIX_CFLAGS_COMPILE = toString (lib.optionals stdenv.cc.isGNU [ "-Wno-error=deprecated-copy" "-Wno-error=redundant-move" "-Wno-error=pessimizing-move" ]);
-
-  inherit nativeBuildInputs buildInputs;
+  # For several reasons building cockroach from source has become
+  # nearly imposible. See https://github.com/NixOS/nixpkgs/pull/152626
+  # Therefore we use the pre-build release binary and wrap it with buildFHSUserEnv to
+  # work on nix.
+  # You can generate the hashes with
+  # nix store prefetch-file <url>
+  src = builtins.fetchurl (lib.getAttr stdenv.system {
```
Comment (wegank) [inline pkgs/servers/sql/cockroachdb/default.nix:16]:
```suggestion
  src = fetchurl (lib.getAttr stdenv.system {
```

Comment (neosimsim) [reply pkgs/servers/sql/cockroachdb/default.nix]: I just noticed that `fetchurl` does not uncompress the tarball. I switched to `fetchzip`. Unfornutately this now results in another hash than `nix store pre-fetch-file` prints.

Comment (neosimsim) [reply pkgs/servers/sql/cockroachdb/default.nix]: `nix flake prefetch` seems to be the correct command.

**Comment by neosimsim** [comment]
I just realized that FHS is obviously not supported for darwin.

**Comment by ankhers** [comment]
Should this be introduced as `cockroachdb-bin` instead?

**Comment by turion** [comment]
> Should this be introduced as cockroachdb-bin instead?

Yes, good point.

**Comment by ankhers** [comment]
Is there any additional movement on this? I would like to get this working now that an upstream bug has been fixed that prevented me from using cockroach.

If the original author does not have time to get this working, I can throw something together that follows this approach.

**Comment by neosimsim** [comment]
Sorry for the delay. I added the change as `cockroachdb-bin` keeping the original `cockroach` as is.

File: pkgs/servers/sql/cockroachdb/cockroachdb-bin.nix [hunk 8]
```diff
@@ -6,21 +6,25 @@
 
 let
   version = "23.1.7";
-  name = "cockroachdb-${version}";
+  name = "cockroachdb";
```
Comment (neosimsim) [inline pkgs/servers/sql/cockroachdb/cockroachdb-bin.nix:9]: @wegank Why did you drop version from the name?

Comment (wegank) [reply pkgs/servers/sql/cockroachdb/cockroachdb-bin.nix:9]: It doesn't seem very intuitive to me that the program name should contain the version. Would you like me to revert the change?

**Review by ankhers — APPROVED** [review]
Approving for now.

I would like to see a generic builder that would allow us to support various versions of cockroach, but that is something we can add at a later point.

**Pull request merged**